### ISSUE TITLE
Bug: Unexpected Verdicts for the Rule "no-control-regex"

### ISSUE THREAD

**Issue by watuwo** [eslint/eslint#15653]
### Environment

Node version: v16.13.0
npm version: v8.1.0
Local ESLint version: v8.9.0 (Currently used)
Global ESLint version: Not found
Operating System: darwin

### What parser are you using?

@typescript-eslint/parser

### What did you do?

<details>
<summary>Configuration</summary>

```
module.exports = {
	env: {
		browser: true,
		es2021: true,
		node: true
	},
	extends: [
		'eslint:recommended',
		'plugin:@typescript-eslint/recommended'
	],
	parser: '@typescript-eslint/parser',
	parserOptions: {
		ecmaVersion: 'latest',
		sourceType: 'module'
	},
	plugins: [
		'@typescript-eslint'
	],
	rules: {
	}
}
```
</details>

```js
const a = new RegExp('\t');
  // expected: error; actual: error
const b = new RegExp('\\t');
  // expected: no error; actual: no error
const c = new RegExp('\u0009');
  // expected: error; actual: error
const d = new RegExp('\\u0009');
  // expected: no error; actual: error
const e = new RegExp('\x09');
  // expected: error; actual: error
const f = new RegExp('\\x09');
  // expected: no error; actual: error
```

### What did you expect to happen?

I expected no errors for the lines defining `d` and `f`. The comments in the reproduction example show my expectations and the observed behavior.

### What actually happened?

I am seeing unexpected errors for the lines defining `d` and `f` (lines 7 and 11):

```
   1:22  error  Unexpected control character(s) in regular expression: \x09  no-control-regex
   5:22  error  Unexpected control character(s) in regular expression: \x09  no-control-regex
   7:22  error  Unexpected control character(s) in regular expression: \x09  no-control-regex
   9:22  error  Unexpected control character(s) in regular expression: \x09  no-control-regex
  11:22  error  Unexpected control character(s) in regular expression: \x09  no-control-regex
```

The errors for the lines 1, 5 and 9 are expected.

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

As a rationale of  the rule https://eslint.org/docs/rules/no-control-regex says: "These characters are rarely used in JavaScript strings so a regular expression containing these characters is most likely a mistake".

Note that the strings `'\\u0009'` and `'\\x09'` do not contain control characters: `'\\u0009' === String.fromCharCode(0x5C, 0x75, 0x30, 0x30, 0x30, 0x39)` and `'\\x09' === String.fromCharCode(0x5C, 0x78, 0x30, 0x39)`.

**Comment by mdjermanovic** [comment]
Hi @watuwo, thanks for the issue!

This works as intended. This rule reports hex and unicode escape sequences as well as raw characters that represent control characters.

> As a rationale of the rule https://eslint.org/docs/rules/no-control-regex says: "These characters are rarely used in JavaScript strings so a regular expression containing these characters is most likely a mistake".

By "strings" it means strings on which the regular expression will be used, not regex pattern strings. The rationale is that any of \u0000-\u001F is more likely a typo than an intention to match that character because most of those characters rarely appear in strings, while those that are common have shorter escape sequences like `\t` and `\n`.

We should clarify this in the documentation, I'll work on that.

**Comment by watuwo** [comment]
Thanks for the feedback and the clarification. Adapting the documentation will let people enable/disable the rule with more confidence.

*Personally* I would have preferred the rule to work according to my initial expectations. I feel like accidentally dropping a backslash is more likely something I would do than accidentally matching a control character. I can understand that other people have other use cases / priorities / deamons. I can understand if you don't want to change the eslint code for that.

When adapting the documentation: there is also a short explanation of the rule at https://eslint.org/docs/rules/ ("disallow control characters in regular expressions"). I don't know whether this should be changed (and if yes, to what).

**Comment by mdjermanovic** [comment]
> I feel like accidentally dropping a backslash is more likely something I would do than accidentally matching a control character. 

Dropping that backslash is not a bug. It results in a pattern with the character's code point, but it works the same as the pattern with an escape sequence for that character.

```js
new RegExp('\t').test("\t"); // true
new RegExp('\\t').test("\t"); // true
new RegExp('\u0009').test("\t"); // true
new RegExp('\\u0009').test("\t"); // true
new RegExp('\x09').test("\t"); // true
new RegExp('\\x09').test("\t"); // true
```

What doesn't look consistent in the behavior of this rule is reporting `new RegExp('\t')`. It's due to intent to report code such as `new RegExp('<TAB>')`, but since the rule only examines the pattern (string value) and doesn't map the pattern back to the original string literal, I believe the reporting of `new RegExp('\t')` is more an implementation side effect than an intent.

**Comment by watuwo** [comment]
I agree, these examples are not errors. It would just have been a matter of personal taste (beeing consistent with `new RegExp('\\x2A')` where `new RegExp('\x2A')` would be an error). Of course `'\\x2A'` has nothing to do with control characters and I would not expect this rule to help me here.

Now knowing that `'\\u0009'` and `'\\x09'` were rejected on purpose, I have deactivated the rule. I am perfectly fine with that.
